### PR TITLE
Clean up RenderingServer and its bindings

### LINKED ISSUE #18621: Directional Shadow Atlas Size Cannot be Modified at Runtime
version 3.0.2 stable

There is currently no way (AFAIK) to change the resolution of the directional shadow atlas from GDScript, which makes changing graphics settings during runtime difficult, since the directional shadow is used in most large 3D scenes.

Additionally, changing the size of the directional shadow atlas in the Project Settings does not actually affect the size of the atlas until Godot is restarted (when it re-reads the godot.project file and properly sets up the directional shadow atlas). I've tried deleting the node and re-adding it, closing and opening the scene, and even creating a new scene, but the shadow atlas size stays the same as when Godot was launched.

### PR DESCRIPTION
**Pull request by reduz**
* Rewrote bindings for RenderingServer.
* They are now all up to date.
* Several unused methods and deprecated features were cleaned up.

*Bugsquad edit: This closes https://github.com/godotengine/godot/issues/18621 and closes https://github.com/godotengine/godot/pull/49762 by superseding it.*

### CODE REVIEW

File: servers/rendering_server.cpp [hunk 1]
```diff
@@ -1473,25 +1694,108 @@ void RenderingServer::_bind_methods() {
...
 	ClassDB::bind_method(D_METHOD("mesh_create"), &RenderingServer::mesh_create);
 	ClassDB::bind_method(D_METHOD("mesh_surface_get_format_offset", "format", "vertex_count", "array_index"), &RenderingServer::mesh_surface_get_format_offset);
 	ClassDB::bind_method(D_METHOD("mesh_surface_get_format_vertex_stride", "format", "vertex_count"), &RenderingServer::mesh_surface_get_format_vertex_stride);
 	ClassDB::bind_method(D_METHOD("mesh_surface_get_format_attribute_stride", "format", "vertex_count"), &RenderingServer::mesh_surface_get_format_attribute_stride);
 	ClassDB::bind_method(D_METHOD("mesh_surface_get_format_skin_stride", "format", "vertex_count"), &RenderingServer::mesh_surface_get_format_skin_stride);
-	//ClassDB::bind_method(D_METHOD("mesh_add_surface_from_arrays", "mesh", "primitive", "arrays", "blend_shapes", "lods", "compress_format"), &RenderingServer::mesh_add_surface_from_arrays, DEFVAL(Array()), DEFVAL(Dictionary()), DEFVAL(ARRAY_COMPRESS_DEFAULT));
+	ClassDB::bind_method(D_METHOD("mesh_add_surface", "mesh", "surface"), &RenderingServer::mesh_add_surface_from_arrays, &RenderingServer::_mesh_add_surface);
```
Comment (akien-mga) [inline servers/rendering_server.cpp:1708]:
```suggestion
	ClassDB::bind_method(D_METHOD("mesh_add_surface", "mesh", "surface"), &RenderingServer::_mesh_add_surface);
```

Not sure how this even compiled.

File: doc/classes/RenderingServer.xml [hunk 2]
```diff
@@ -3106,16 +4873,12 @@
...
-			Blend shapes are normalized.
 		</constant>
 		<constant name="BLEND_SHAPE_MODE_RELATIVE" value="1" enum="BlendShapeMode">
-			Blend shapes are relative to base weight.
 		</constant>
 		<constant name="MULTIMESH_TRANSFORM_2D" value="0" enum="MultimeshTransformFormat">
-			Use [Transform2D] to store MultiMesh transform.
 		</constant>
 		<constant name="MULTIMESH_TRANSFORM_3D" value="1" enum="MultimeshTransformFormat">
-			Use [Transform3D] to store MultiMesh transform.
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:3118]: These descriptions were lost in the process, could be readded.

File: doc/classes/RenderingServer.xml [hunk 3]
```diff
@@ -3579,53 +5476,51 @@
...
-			Uses the default filter mode for this [Viewport].
 		</constant>
 		<constant name="CANVAS_ITEM_TEXTURE_FILTER_NEAREST" value="1" enum="CanvasItemTextureFilter">
-			The texture filter reads from the nearest pixel only. The simplest and fastest method of filtering, but the texture will look pixelized.
 		</constant>
 		<constant name="CANVAS_ITEM_TEXTURE_FILTER_LINEAR" value="2" enum="CanvasItemTextureFilter">
-			The texture filter blends between the nearest 4 pixels. Use this when you want to avoid a pixelated style, but do not want mipmaps.
 		</constant>
 		<constant name="CANVAS_ITEM_TEXTURE_FILTER_NEAREST_WITH_MIPMAPS" value="3" enum="CanvasItemTextureFilter">
-			The texture filter reads from the nearest pixel in the nearest mipmap. The fastest way to read from textures with mipmaps.
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:3601]: Also lost a bunch here.

File: servers/rendering/renderer_rd/renderer_scene_gi_rd.h [hunk 4]
```diff
@@ -611,9 +611,9 @@ class RendererSceneGIRD {
 		uint32_t blend_ambient;
 		uint32_t texture_slot;
 
-		float anisotropy_strength;
-		float ao;
-		float ao_size;
+		uint32_t pad0;
+		uint32_t pad1;
+		uint32_t pad2;
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/renderer_scene_gi_rd.h:616]: Is this just to pad to a specific size? Would be good to add a comment to make it explicit, so we know what to remove if adding more fields and reclaiming the padding.

File: servers/rendering/renderer_viewport.h [hunk 5]
```diff
@@ -68,8 +68,7 @@ class RendererViewport {
 		Rect2 viewport_to_screen_rect;
 		bool viewport_render_direct_to_screen;
 
-		bool hide_scenario;
-		bool hide_canvas;
+		bool disable_2d;
```
Comment (akien-mga) [inline servers/rendering/renderer_viewport.h:71]: I guess to be consistent `disable_3d` should also be left uninitialized here and initialized when creating a struct.

File: servers/rendering/rendering_server_default.cpp [hunk 6]
```diff
@@ -42,26 +42,6 @@
 
 int RenderingServerDefault::changes = 0;
 
-/* BLACK BARS */
```
Comment (akien-mga) [inline servers/rendering/rendering_server_default.cpp:45]: Is this possible through a different API?
The UX of this method was maybe not the best, but I think being able to customize black bars is still something that some users enjoy, so we should re-add something for it eventually.

File: servers/rendering_server.cpp [hunk 7]
```diff
@@ -1511,15 +1815,22 @@ void RenderingServer::_bind_methods() {
...
 	ClassDB::bind_method(D_METHOD("skeleton_get_bone_count", "skeleton"), &RenderingServer::skeleton_get_bone_count);
 	ClassDB::bind_method(D_METHOD("skeleton_bone_set_transform", "skeleton", "bone", "transform"), &RenderingServer::skeleton_bone_set_transform);
 	ClassDB::bind_method(D_METHOD("skeleton_bone_get_transform", "skeleton", "bone"), &RenderingServer::skeleton_bone_get_transform);
 	ClassDB::bind_method(D_METHOD("skeleton_bone_set_transform_2d", "skeleton", "bone", "transform"), &RenderingServer::skeleton_bone_set_transform_2d);
 	ClassDB::bind_method(D_METHOD("skeleton_bone_get_transform_2d", "skeleton", "bone"), &RenderingServer::skeleton_bone_get_transform_2d);
+	ClassDB::bind_method(D_METHOD("skeleton_set_base_transform_2d", "skeleton", "base_transform"), &RenderingServer::skeleton_set_base_transform_2d);
+
+	/* Light API */
 
-#ifndef _3D_DISABLED
```
Comment (akien-mga) [inline servers/rendering_server.cpp:1522]: Did you check that `disable_3d=yes` still works?

File: doc/classes/RenderingServer.xml [hunk 8]
```diff
@@ -155,2764 +205,4457 @@
...
 			<description>
-				Sets the world space transform of the instance. Equivalent to [member Node3D.transform].
 			</description>
 		</method>
-		<method name="instance_set_visibility_parent">
+		<method name="occluder_set_mesh">
 			<return type="void">
 			</return>
-			<argument index="0" name="instance" type="RID">
+			<argument index="0" name="arg0" type="RID">
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:2929]: Missing argument names for `occluder_set_mesh`.

File: doc/classes/RenderingServer.xml [hunk 9]
```diff
@@ -155,2764 +205,4457 @@
...
 		</method>
-		<method name="scenario_set_environment">
+		<method name="texture_set_force_redraw_if_visible">
 			<return type="void">
 			</return>
-			<argument index="0" name="scenario" type="RID">
+			<argument index="0" name="texture" type="RID">
 			</argument>
-			<argument index="1" name="environment" type="RID">
+			<argument index="1" name="arg1" type="bool">
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:4022]: Missing argument name for `texture_set_force_redraw_if_visible`.

File: doc/classes/RenderingServer.xml [hunk 10]
```diff
@@ -155,2764 +205,4457 @@
...
+			<return type="void">
 			</return>
+			<argument index="0" name="decal" type="RID">
+			</argument>
+			<argument index="1" name="distance" type="bool">
+			</argument>
+			<argument index="2" name="arg2" type="float">
+			</argument>
+			<argument index="3" name="arg3" type="float">
+			</argument>
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:1144]: Missing argument names for `decal_set_distance_fade`.

File: doc/classes/RenderingServer.xml [hunk 11]
```diff
@@ -155,2764 +205,4457 @@
...
 			<return type="void">
 			</return>
-			<argument index="0" name="instance" type="RID">
+			<argument index="0" name="decal" type="RID">
 			</argument>
-			<argument index="1" name="flag" type="int" enum="RenderingServer.InstanceFlags">
+			<argument index="1" name="fade" type="float">
 			</argument>
-			<argument index="2" name="enabled" type="bool">
+			<argument index="2" name="arg2" type="float">
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:1175]: Missing arg name for `decal_set_fade`.

File: doc/classes/RenderingServer.xml [hunk 12]
```diff
@@ -155,2764 +205,4457 @@
...
+			</description>
+		</method>
+		<method name="environment_bake_panorama">
+			<return type="Image">
+			</return>
+			<argument index="0" name="bake_irradiance" type="RID">
+			</argument>
+			<argument index="1" name="size" type="bool">
+			</argument>
+			<argument index="2" name="arg2" type="Vector2i">
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:1246]: Missing arg name for `environment_bake_panorama`.

File: doc/classes/RenderingServer.xml [hunk 13]
```diff
@@ -18,42 +18,92 @@
...
-			<argument index="3" name="bottom" type="int">
+			<argument index="3" name="far_transition" type="float">
+			</argument>
+			<argument index="4" name="near_enable" type="bool">
+			</argument>
+			<argument index="5" name="near_distance" type="float">
+			</argument>
+			<argument index="6" name="near_transition" type="float">
+			</argument>
+			<argument index="7" name="arg7" type="float">
```
Comment (akien-mga) [inline doc/classes/RenderingServer.xml:76]: Missing arg name for `camera_effects_set_dof_blur`.

File: servers/rendering/renderer_rd/renderer_scene_gi_rd.h (hunk 4 above)
Comment (reduz) [reply servers/rendering/renderer_rd/renderer_scene_gi_rd.h:616]: its a push constant for Vulkan, needs to be paded to 16, but there are lots of those like this around the renderer.

File: servers/rendering/rendering_server_default.cpp (hunk 6 above)
Comment (reduz) [reply servers/rendering/rendering_server_default.cpp:45]: It needs to be re-added again at some point, but with a different API and per window (not the case right now).

File: servers/rendering_server.cpp (hunk 7 above)
Comment (reduz) [reply servers/rendering_server.cpp:1522]: It works, but in general this needs a makeover to be done properly, including not compiling the actual 3D renderers, shaders, etc. and use the dummies instead.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by reduz** [comment]
Hooray!

**Pull request merged**